File: Ruby/Civitas/pkg/Civitas-0.0.1/lib/jugador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

# encoding: utf-8
require_relative 'tablero'
require_relative 'diario'
require_relative 'dado'
require_relative 'titulo_propiedad'
require_relative 'sorpresa'

module Civitas
  class Jugador
    def initialize(nombre)
      @nombre=nombre
      @encarcelado=false
      @num_casilla_actual=0
      @puede_comprar=false
      @@Casas_max=4
      @@Casas_por_hotel=4
      @@Hoteles_max=4
      @@Paso_por_salida=1000
      @@Precio_libertad=200
      @@Saldo_inicial=7500
      @saldo=@@Saldo_inicial
      
      @propiedades=[]
      @salvoconducto
    end
    
    attr_reader  :Casas_max, :Casas_por_hotel, :Hoteles_max, :Paso_por_salida, :Precio_libertad, :Saldo_inicial
    attr_accessor :nombre, :encarcelado, :num_casilla_actual, :puede_comprar, :saldo, :propiedades, :salvoconducto
    
    protected
    
    def copia(otro)
      @nombre=otro.nombre
      @encarcelado=otro.encarcelado
      @num_casilla_actual=otro.num_casilla_actual
      @puede_comprar=otro.puede_cpmprar
      @saldo=otro.saldo
    end
    
    public
    
    def debe_ser_encarcelado
      debe=false
      
      if(!encarcelado)
        if(!tiene_salvoconducto)
          debe=true
          mensaje="El jugador debe ser encarcelado"
          Diario.instance.ocurre_evento(mensaje)
        else
          perder_salvoconducto
          debe=false
          mensaje="El jugador se libra de ser encarcelado"
          Diario.instance.ocurre_evento(mensaje)
        end
      end
      
      return debe
    end
    
    def get_nombre
      nom=@nombre
      
      return nom
    end
    
    def encarcelar(num_casilla_carcel)
      encarcelar=@encarcelado
      
      if(debe_ser_encarcelado==true && @num_casilla_actual==num_casilla_carcel)
        self.mover_a_casilla(num_casilla_carcel)
        @encarcelado=true
        encarcelar=@encarcelado
        mensaje="Jugador encarcelado"
        Diario.instance.ocurre_evento(mensaje)
      end
      
      return encarcelar
    end
    
    def obtener_salvoconducto(s)
      obtener=false
      
      if(!encarcelado)
        @salvoconducto=s
        obtener=true
      end
      
      return obtener
    end
    
    def perder_salvoconducto
      @salvoconducto.usada()
      @salvoconducto=nil
    end
    
    def tiene_salvoconducto
      tiene=false
      
      if(salvoconducto!=nil)
        tiene=true
      end
      
      return tiene
    end
    
    def puede_comprar_casilla
      puede=false
      
      if(encarcelado)
        @puede_comprar=false
        puede=@puede_comprar
      else
        @puede_comprar=true
        puede=@puede_comprar
      end
      
      return puede
    end
    
    def paga(cantidad)
      paga=modificar_saldo(cantidad*(-1))
      return paga
    end
    
    def paga_impuesto(cantidad)
      paga=false
      
      if(!encarcelado)
        paga=paga(cantidad)
      end
      
      return paga
    end
    
    def paga_alguiler
      paga=false
      
      if(!encarcelado)
        paga=paga(cantidad)
      end
      
      return paga
    end
    
    def recibe(cantidad)
      recibe=false
      
      if(!encarcelado)
        recibe=modificar_saldo(cantidad)
      end
      
      return recibe
    end
    
    def modificar_saldo(cantidad)
      @saldo=@saldo+cantidad
      mensaje="Saldo modificado correctamente, nuevo saldo -> #{@saldo}"
      Diario.instance.ocurre_evento(mensaje)
      return true
    end
    
    def get_propiedades
      p=@propiedades
      
      return p
    end
    
    public
    
    def get_saldo
      s=@saldo
      
      return s
    end
    
    def mover_a_casilla(num_casilla)
      mover=false
      
      if(!encarcelado)
        @num_casilla_actual=num_casilla
        @puede_comprar=false
        mensaje="\nEl jugador avanza a #{num_casilla}"
        Diario.instance.ocurre_evento(mensaje)
        mover=true
      end
      
      return mover
    end
    
    def get_num_casilla_actual
      casilla=@num_casilla_actual
      
      return casilla
    end
    
    def get_encarcelado
      encar=@encarcelado
      
      return encar
    end
    
    def puedo_gastar(precio)
      puedo=false
      
      if(!encarcelado && saldo>=precio)
        puedo=true
      end
      
      return puedo
    end
    
    def comprar(titulo)
      comprar=false
      
      if(!encarcelado && puede_comprar)
          precio=titulo.get_precio_compra
        if(puedo_gastar(precio))
            comprar=titulo.comprar(self)
           if(comprar)
             @propiedades.push(titulo)
             Diario.instance.ocurre_evento("\n#{@nombre} ha comprado #{titulo.get_nombre}")
           end
        end
      end
      
      return comprar
    end
    
    def vender(ip)
      vender=false
      
      if(!encarcelado && existe_la_propiedad(ip))
        @propiedades.delete_at(ip)
        vender=true
      end
      
      return vender
    end
    
    def hipotecar(ip)
      hipotecar=false
      
      if(!encarcelado && existe_la_propiedad(ip))
        propiedad=@propiedades[ip]
        hipotecar=propiedad.hipotecar(self)
      end
      
      if(hipotecar)
        Diario.instance.ocurre_evento("#{@nombre} Ha hipotecado la propiedad")
      end
      
      return hipotecar
    end
    
    def cantidad_casas_hoteles # Revisar 
      ch=0
      
      for i in (0..@propiedades.length)
        ch+=@propiedades[i].cantidad_casas_hoteles
      end
      
      return ch
    end
    
    def cancelar_hipoteca(ip)
      cancelar=false
      
      if(existe_la_propiedad(ip) && !encarcelado)
        propiedad=@propiedades[ip]
        cantidad=propiedad.get_importe_cancelar_hipoteca
        puedo_gastar=puedo_gastar(cantidad)
        
        if(puedo_gastar)
          cancelar=propiedad.cancelar_hipoteca(self)
        end
        
        if(cancelar)
          Diario.instance.ocurre_evento("#{@nombre} Ha cancelado la hipoteca")
        end
      end
      
      return cancelar
    end
    
    def en_banca_rota
      ebr=false
      
      if(saldo<1)
        ebr=true
      end
      
      return ebr
    end
    
    def tiene_algo_que_gestionar
      tiene=false
      
      if(propiedades.length>0)
        tiene=true
      end
      
      return tiene
    end
    
    def puede_salir_carcel_pagando
      puede=false
      
      if(saldo>get_precio_libertad)
        puede=true
      end
      
      return puede
    end
    
    def salir_carcel_pagando
      salir=false
      
      if(encarcelado && puede_salir_carcel_pagando)
        self.paga(get_precio_libertad)
        @encarcelado=false
        mensaje="El jugador "+@nombre+" ha salido de la carcel pagando"
        Diario.instance.ocurre_evento(mensaje)
        salir=true
      end
      
      return salir
    end
    
    def salir_carcel_tirando
      salir=false
      
      if(Dado.instance.salgo_de_la_carcel)
        salir=true
        mensaje="El jugador "+@nombre+" ha salido de la carcel tirando"
        Diario.instance.ocurre_evento(mensaje)      
      end
      
      return salir
    end
    
    private 
    
    def existe_la_propiedad(ip)
      existe=false
      
      for i in (0..@propiedades.length)
        if(propiedades[i]==propiedades[ip])
          existe=true
        end
      end
      
      return existe
    end
    
    def get_premio_paso_salida
      premio=@@Paso_por_salida
      
      return premio
    end
    
    def get_precio_libertad
      precio=@@Precio_libertad
      
      return precio
    end
    
    public
    
    def pasa_por_salida
      modificar_saldo(get_premio_paso_salida)
      mensaje="El jugador "+@nombre+" ha pasado por salida y ha obtenido "+@@Paso_por_salida+" euros"
      Diario.instance.ocurre_evento(mensaje)
      return true
    end
    
    def puedo_edificar_casa(propiedad)
      existe=false
      puedo=false
      
      i=0
      while(i<@propiedades.length)
        if(@propiedades[i].get_propietario==@propiedades[propiedad].get_propietario)
          existe=true
        end
        i+=1
      end
      
      if(existe && propiedad.get_num_casas<@@Casas_max)
        puedo=true
      end
      
      return puedo
    end
    
    def puedo_edificar_hotel(propiedad)
      existe=false
      puedo=false
      
      for i in (0..@propiedades.length)
        if(@propiedades[i].get_propietario==@propiedades[propiedad].get_propietario)
          existe=true
        end
      end
      
      if(existe && propiedad.get_num_hoteles<@@Hoteles_max)
        puedo=true
      end
      
      return puedo
    end
    
    def construir_casa(ip)
      construir=false
      
      p=@propiedades[ip]
      puedo=puedo_edificar_casa(p)
      precio=p.get_precio_edificar
      
      if(puedo_gastar(precio && puedo_edificar_casa(ip)))
        puedo=true
      end
      if(puedo)
        construir=true
        Diario.instance.ocurre_evento("\n#{@nombre} ha construido una casa eh #{p}")
      end
      
      return construir
    end
    
    def construir_hotel(ip)
      construir=false
      
      p=@propiedades[ip]
      puedo=puedo_edificar_hotel(p)
      precio=p.get_precio_edificar
      
      if(puedo_gastar(precio && puedo_edificar_hotel(ip)))
        puedo=true
      end
      if(puedo)
        construir=true
        Diario.instance.ocurre_evento("\n#{@nombre} ha construido un hotel eh #{p}")
      end
      
      return construir
    end
    
    def puedo_edificar_hotel(titulo)
      puedo=false
      precio=titulo.get_precio_edificar
      
      if(puedo_gastar(precio) && titulo.get_num_hoteles<@@Hoteles_max && titulo.get_num_casas>=@@Casas_por_hotel)
        puedo=true
      end
      
      return puedo
    end
    
    def construir_hotel(ip)
      construir=false
      
      if(!encarcelado && existe_la_propiedad(ip))
        propiedad=@propiedades[ip]
        puedo=puedo_edificar_hotel(propiedad)
        precio=propiedad.get_precio_edificar
        
        if(puedo_gastar(precio) && propiedad.get_num_hoteles<@@Hoteles_max && propiedades.get_num_casas>=@@Casas_por_hotel)
          puedo=true
        end
        
        if(puedo)
          construir=propiedad.construir_hotel(self)
          propiedad.derruir_casas(@@Casas_por_hotel, self)
        end
        
        Diario.instance.ocurre_evento("#{@nombre} Ha construido un hotel")
      end
      
      return construir
    end
    
    def compare_to(otro)
      return self.saldo <=> otro.saldo
      # Si el saldo del primer jugador es menor, devuelve -1
      # Si el saldo de ambos jugadores es igual, devuelve 0
      # Si el saldo del segundo jugador es menor, devuelve 1
    end
    
    def to_string
      puts "Nombre -> #{@nombre}
            Encarcelado -> #{@encarcelado}
            Saldo -> #{@saldo}
            Casilla actual -> #{@num_casilla_actual}"
    end
  end
end

File: Java/P4/Civitas/src/civitas/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author arturo
 */
public class Jugador implements Comparable<Jugador>{
    protected static int CasasMax=4;
    protected static int PorHotel=4;
    protected boolean encarcelado;
    protected static int HotelesMax=4;
    private String nombre;
    private int numCasillaActual;
    protected static float PasoPorSalida=1000;
    protected static float PrecioLibertad=200;
    private boolean puedeComprar;
    private float saldo;
    static private float SaldoInicial=7500;
    private ArrayList<tituloPropiedad> propiedades;
    Dado nuevo=Dado.getInstance();
    private Sorpresa salvoconducto=null;
    
    Jugador(String nombre){
        saldo=SaldoInicial;
        this.nombre=nombre;
        puedeComprar=true;
        numCasillaActual=0;
        propiedades=new ArrayList<>();
    }
    
    protected Jugador(Jugador otro){
        encarcelado=otro.estaEncarcelado();
        nombre=otro.getNombre();
        numCasillaActual=otro.getNumCasillaActual();
        puedeComprar=otro.getPuedeComprar();
        saldo=otro.getSaldo();
        salvoconducto=otro.getSalvoconducto();
        propiedades=otro.getPropiedades();
    }
    
    public boolean estaEncarcelado(){
        return encarcelado;
    }
    
    public void setPuedeComprar(boolean x){
        this.puedeComprar=x;
    }
    
    boolean getPuedeComprar(){
        return puedeComprar;
    }
    
    public float getSaldo(){
        return saldo;
    }
    
    Sorpresa getSalvoconducto(){
        return salvoconducto;
    }
    
    public String getNombre(){
        return nombre;
    }
    
    public int getNumCasillaActual(){
        return this.numCasillaActual;
    }
    
    int getCasasMax(){
        return CasasMax;
    }
    
    int getHotelesMax(){
        return HotelesMax;
    }
    
    public ArrayList<tituloPropiedad> getPropiedades(){
        return propiedades;
    }
    
    boolean enBancaRota(){
        boolean aux=false;
        
        if(this.saldo<=0){
            aux=true;
        }
        
        return aux;
    }
    
    protected boolean debeSerEncarcelado(){
        boolean aux=false;
        
        if(tieneSalvoconducto()==false && encarcelado==false){
            aux=true;
        }else{
            Diario.getInstance().ocurreEvento("El jugador se libra de la cárcel y pierde el salvoconducto\n");
            perderSalvoconducto();
            aux=false;
        }
        
        return aux;
    }
    
    int cantidadCasasHoteles(){
        int total=0;
        
        for(int i=0; i<propiedades.size(); i++){
            total+=propiedades.get(i).getNumCasas();
            total+=propiedades.get(i).getNumHoteles();
        }
        
        return total;
    }
    
    boolean encarcelar(int numCasillaCarcel){
        if(debeSerEncarcelado()){
            Diario.getInstance().ocurreEvento("El jugador es encarcelado\n");
            this.moverACasilla(numCasillaCarcel);
            this.encarcelado=true;
        }
        
        return true;
    }
    
    boolean obtenerSalvoconducto(Sorpresa s){
        boolean aux=false;
        
        if(encarcelado==false){
            this.salvoconducto=s;
            aux=true;
        }
        
        return aux;
    }
    
    void perderSalvoconducto(){
        obtenerSalvoconducto(null);
    }
    
    public boolean esEspeculador(){
        return false;
    }
    
    public boolean tieneSalvoconducto(){
        boolean aux=false;
        
        if(!obtenerSalvoconducto(null)){
            aux= true;
        }
        
        return aux;
    }
    
    boolean puedeComprarCasilla(){
        if(encarcelado){
            puedeComprar=false;
        }else{
            puedeComprar=true;
        }
        
        return puedeComprar;
    }
    
    boolean paga(float cantidad){
        return this.modificarSaldo(cantidad*(-1));
    }
    
    boolean pagaImpuesto(float cantidad){
        boolean aux=false;
        
        if(!encarcelado){
            aux=this.paga(cantidad);
        }
        
        return aux;
    }
    
    boolean pagaAlquiler(float cantidad){
        boolean aux=false;
        
        if(encarcelado==false){
            aux=this.paga(cantidad);
        }
        
        return aux;
    }
    
    boolean recibe(float cantidad){
        boolean aux=false;
        
        if(encarcelado==false){
            aux=this.modificarSaldo(cantidad);
        }
        
        return aux;
    }
    
    boolean modificarSaldo(float cantidad){
        this.saldo+=(float)cantidad;
        Diario.getInstance().ocurreEvento("\nEl saldo del jugador "+this.nombre+" ha sido modificado\n");
        return true;
    }
    
    boolean moverACasilla(int numCasilla){
        boolean aux=false;
        
        if(encarcelado==false){
            this.numCasillaActual=numCasilla;
            this.puedeComprar=false;
            Diario.getInstance().ocurreEvento("\n"+this.nombre+" se ha movido a la casilla: "+this.numCasillaActual+"\n");
            aux=true;
        }
        
        return aux;
    }
    
    private boolean puedoGastar(float precio){
        boolean aux=false;
        
        if(!encarcelado && saldo>=precio){
            aux=true;
        }
        
        return aux;
    }
    
    boolean vender(int ip){
        boolean aux=false;
        
        if(!encarcelado && (existeLaPropiedad(ip)==true && propiedades.get(ip).getHipotecado()==false)){
            tituloPropiedad titulo=propiedades.get(ip);
            aux=titulo.Vender(this);
            
            Diario.getInstance().ocurreEvento(nombre+" ha vendido la casilla "+propiedades.get(ip));
            propiedades.remove(ip);
        }
        
        return aux;
    }
    
    boolean comprar(tituloPropiedad titulo){
        boolean aux=false;
        
        if(encarcelado){
            return aux;
        }
        
        if(puedeComprar==true){
                float precio=titulo.getPrecioVenta();
            if(puedoGastar(precio)==true){
                    aux=titulo.Comprar(this);
                if(aux==true){
                    propiedades.add(titulo);
                    Diario.getInstance().ocurreEvento("El jugador "+this+" compra la propiedad "+titulo.toString());
                    puedeComprar=false;
                }
            }
        }
        return aux;
    }
    
     boolean hipotecar(int ip){
        boolean aux=false;
        
        if(encarcelado==false){
            if(existeLaPropiedad(ip)){
                tituloPropiedad titulo=propiedades.get(ip);
                aux=titulo.hipotecar(this);
            }
        }
        
        if(aux==true){
            Diario.getInstance().ocurreEvento("El jugador "+nombre+" ha hipotecado la propiedad "+ip);
        }
            
        return aux;
    }
    
    boolean cancelarHipoteca(int ip){
        boolean aux=false;
        
        if(!encarcelado){
            if(existeLaPropiedad(ip)){
                tituloPropiedad titulo=propiedades.get(ip);
                float cantidad=titulo.getImporteCancelarHipoteca();
                Boolean puedoGastar=puedoGastar(cantidad);
                
                if(puedoGastar==true){
                        aux=titulo.cancelarHipoteca(this);
                    if(aux==true){
                        Diario.getInstance().ocurreEvento("El jugador "+nombre+" cancela la hipoteca de la propiedad "+ip);
                    }
                }
            }
        }
        
        return aux;
    }
    
    boolean tieneAlgoQueGestionar(){
        boolean aux=false;
        
        if(!propiedades.isEmpty()){
            aux=true;
        }
        
        return aux;
    }
    
    private boolean puedeSalirCarcelPagando(){
        boolean aux=false;
        
        if(encarcelado && saldo>=PrecioLibertad){
            aux=true;
        }
        
        return aux;
    }
    
    boolean salirCarcelPagando(){
        boolean aux=false;
        
        if(encarcelado && puedeSalirCarcelPagando()){
            this.paga(PrecioLibertad);
            encarcelado=false;
            Diario.getInstance().ocurreEvento(this.nombre+" sale de la cárcel pagando");
            aux=true;
        }
        
        return aux;
    }
    
    @Override
    public String toString(){
        String devolver=" ";
        String todas=" ";
        String aux;
        
        for(int i=0; i<propiedades.size(); i++){
            aux=propiedades.get(i).getNombre();
            todas=todas+"";
            todas=todas+aux;
        }
        
        devolver="\n"+this.getNombre()+"\nEncarcelado: "+this.encarcelado
                +"\nPuede comprar: "+this.puedeComprar+"\nSaldo: "+this.saldo+"\nNumero de propiedades: "+
                this.propiedades.size()+"\nPropiedades: "+todas+" "+"\nCasilla actual: "+this.numCasillaActual+"\n-----------------------------------------";
        
        return devolver;
    }
            
    boolean salirCarcelTirando(){
        boolean aux=false;
        
        if(encarcelado && nuevo.SalgoDeLaCarcel()){
            aux=true;
            encarcelado=false;
            Diario.getInstance().ocurreEvento(this.nombre+" sale de la cárcel tirando");
        }
        
        return aux;
    }  
    
    boolean construirCasa(int ip){
        boolean aux=false;
        boolean puedo=false;
        
        if(!encarcelado){
                boolean existe=existeLaPropiedad(ip);
            if(existe){
                    tituloPropiedad propiedad=propiedades.get(ip);
                    puedo=this.puedoEdificarCasa(propiedad);
                if(puedo==true){
                        aux=propiedad.construirCasa(this);
                    if(aux==true){
                        Diario.getInstance().ocurreEvento("El jugador "+nombre+" construye casa en "+propiedad.getNombre()+"\n");
                    }
                }
            }
        }
        return aux;
    }
    
    boolean construirHotel(int ip){
        boolean aux=false;
        boolean puedo=false;
        
        if(!encarcelado){
                boolean existe=existeLaPropiedad(ip);
            if(existe){
                    tituloPropiedad propiedad=propiedades.get(ip);
                    puedo=this.puedoEdificarCasa(propiedad);
                if(puedo==true){
                        aux=propiedad.construirHotel(this);
                    if(aux==true){
                        Diario.getInstance().ocurreEvento("El jugador "+nombre+" construye hotel en "+propiedad.getNombre()+"\n");
                    }else{
                        Diario.getInstance().ocurreEvento("\nEl numero de casas no es suficiente (debe ser mayor a 4)\n");
                    }
                }
            }
        }
        return aux;
    }
            
    boolean pasaPorSalida(){
        modificarSaldo(PasoPorSalida);  
        Diario.getInstance().ocurreEvento(this.nombre+" pasa por la casilla de salida y recibe el pago correspondiente");
        return true;
    }
    
    private boolean existeLaPropiedad(int ip){
        return true;
    }
    
    private boolean puedoEdificarCasa(tituloPropiedad casa){
        boolean aux=false;
        float precio=casa.getPrecioEdificar();
        
        if(this.puedoGastar(precio) && casa.getNumCasas()<this.getCasasMax()){
            aux=true;
        }
        
        return aux;
    }
            
    /**
     *
     * @param otro
     * @return
     */
    @Override
    public int compareTo(Jugador otro){
        int aux;
        
        if(this.saldo>=otro.saldo){
            aux=(int) this.saldo-(int) otro.saldo;
        }else{
            aux=(int) otro.saldo-(int) this.saldo;
        }
        
        return aux;
    }       
}